File: mean-free-path.cpp
// -----------------------------------------------------------------------------
// Feynman Lectures - Volume I, Chapter 1, Exercise 1-9
// Estimate of Mean Free Path for Air Molecules at STP (with CLI input)
//
// Author: Scott K. Douglass
// Date: 2025-07-20
// Description:
//     Computes mean free path using:
//         λ = 1 / (√2 * π * d² * n)
//     Allows user to press Enter to use default values.
//
// Dependencies: <iostream>, <cmath>, <limits>
// -----------------------------------------------------------------------------

#include <cmath>
#include <iostream>
#include <limits>
#include <string>

double get_input_with_default(const std::string &prompt, double default_value) {
  std::string input;
  std::cout << prompt << " [default: " << default_value << "]: ";
  std::getline(std::cin, input);
  try {
    return input.empty() ? default_value : std::stod(input);
  } catch (...) {
    std::cerr << "Invalid input. Using default.\n";
    return default_value;
  }
}

int main() {
  const double default_d = 3.7e-10; // meters
  const double default_n = 2.7e25;  // m^-3
  const double pi = 3.14159265358979323846;

  double d =
      get_input_with_default("Enter molecular diameter (meters)", default_d);
  double n = get_input_with_default("Enter number density (m^-3)", default_n);

  double lambda = 1.0 / (std::sqrt(2.0) * pi * d * d * n);

  std::cout << "\nEstimated mean free path ≈ " << lambda << " meters\n";

  return 0;
}

File: mean-free-path.py
# ------------------------------------------------------------------------------
# Estimate of Mean Free Path for Air Molecules at STP
#
# Author: Scott K. Douglass
# Date: 2025-07-20
# Description:
#     Computes the mean free path of air molecules at standard temperature
#     and pressure (STP) using classical kinetic theory:
#
#         λ = 1 / (√2 * π * d² * n)
#
#     where:
#         d = molecular diameter (approx. 3.7 × 10⁻¹⁰ m)
#         n = number density of air molecules (~2.7 × 10²⁵ m⁻³)
#
# Dependencies: math
# ------------------------------------------------------------------------------


import math

# Default values
DEFAULT_DIAMETER = 3.7e-10  # meters
DEFAULT_DENSITY = 2.7e25    # particles per m³

# Prompt user for input (or use defaults)
try:
    d_input = input(f"Enter molecular diameter in meters [default: {DEFAULT_DIAMETER:.2e}]: ").strip()
    d = float(d_input) if d_input else DEFAULT_DIAMETER

    n_input = input(f"Enter number density in m⁻³ [default: {DEFAULT_DENSITY:.2e}]: ").strip()
    n = float(n_input) if n_input else DEFAULT_DENSITY

    lambda_mfp = 1 / (math.sqrt(2) * math.pi * d**2 * n)

    print(f"\nEstimated mean free path ≈ {lambda_mfp:.2e} meters")

except ValueError:
    print("Invalid input. Please enter valid numbers or press Enter for default.")